compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt

```diff
@@ -428,6 +453,129 @@ class IncrementalContext(
         updateOutputs(outputs, cleanOutputs)
 
     }
+
+    fun recordLookup(psiFile: PsiJavaFile, fqn: String) {
+        val path = psiFile.virtualFile.path
+        val name = fqn.substringAfterLast('.')
+        val scope = fqn.substringBeforeLast('.', "<anonymous>")
+
+        fun record(scope: String, name: String) =
+            lookupTracker.record(path, Position.NO_POSITION, scope, ScopeKind.CLASSIFIER, name)
+
+        record(scope, name)
+        val onDemandImports =
+                psiFile.getOnDemandImports(false, false).mapNotNull { (it as? PsiPackage)?.qualifiedName }
+        if (scope in onDemandImports) {
+            record(psiFile.packageName, name)
+            onDemandImports.forEach {
+                record(it, name)
+            }
+        }
+    }
+
+    private fun recordLookup(ref: PsiClassReferenceType, def: PsiClass) {
+        val psiFile = ref.reference.containingFile as? PsiJavaFile ?: return
+        // A type parameter doesn't have qualified name.
+        def.qualifiedName?.let { recordLookup(psiFile, it) }
+    }
+
+    fun recordLookup(ref: PsiType) {
+        when (ref) {
+            is PsiArrayType -> recordLookup(ref.componentType)
+            is PsiClassReferenceType -> {
+                val def = ref.resolve() ?: return
+                recordLookup(ref, def)
+                // in case the corresponding KotlinType is passed through ways other than KSTypeReferenceJavaImpl
+                ref.typeArguments().forEach {
+                    if (it is PsiType) {
+                        recordLookup(it)
+                    }
+                }
+            }
+            is PsiWildcardType -> ref.bound?.let { recordLookup(it) }
+        }
+    }
+
+    fun recordLookupWithSupertypes(kotlinType: KotlinType) {
+        (listOf(kotlinType) + kotlinType.supertypes()).mapNotNull {
+            it.constructor.declarationDescriptor?.findPsi() as? PsiClass
+        }.forEach {
+            it.superTypes.forEach {
+                recordLookup(it)
+            }
+        }
+    }
+
+    private fun recordLookup(psi: PsiField) {
+        recordLookup(psi.type)
+    }
+
+    private fun recordLookup(psi: PsiMethod) {
+        psi.parameterList.parameters.forEach {
+            recordLookup(it.type)
+        }
+        psi.returnType?.let { recordLookup(it) }
+        psi.typeParameters.forEach {
+            it.bounds.mapNotNull { it as? PsiType }.forEach {
+                recordLookup(it)
+            }
+        }
+    }
+
+    fun recordLookupForDeclaration(declaration: KSDeclaration) {
+        when (declaration) {
+            is KSPropertyDeclarationJavaImpl -> recordLookup(declaration.psi)
+            is KSFunctionDeclarationJavaImpl -> recordLookup(declaration.psi)
+        }
+    }
+
+    fun recordLookupForCallableMemberDescriptor(descriptor: CallableMemberDescriptor) {
+        val psi = descriptor.findPsi()
+        when (psi) {
+            is PsiMethod -> recordLookup(psi)
```

Comment:
what about just make `recordLookup` take care of type checking?

Reply:
Can you give an example?

Reply:
make `recordLookup(psi: PsiElement)` and then handle psi type checking there.

Reply:
Will rename to recordLookupPsiMethod() and recordLookupPsiField().